activestorage/lib/active_storage/service/s3_service.rb

```diff
@@ -14,6 +14,18 @@ class Service::S3Service < Service
     attr_reader :client, :bucket
     attr_reader :multipart_upload_threshold, :upload_options
 
+    class << self
+      def options_from_uri(uri:, query_hash: {}) # :nodoc:
```

Comment:
There's something strange about parsing this query_hash around like this. Almost like we need a `ServiceUrl` object that encapsulates this. Don't like it otherwise. And it also just seems weirdly magic with no clarity (or documentation) on what can be in it.